### ISSUE TITLE
Should `isinstance(f, Callable)` and `if callable(f)` behave the same?

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#4203]
With the new gradual `isinstance` narrowing behavior, we do relax the behavior of `isinstance(f, Callable)`, but we do not change the result of `if callable(f)`, which remains strict:

```py
from typing import Callable, reveal_type

def _(f: object):
    if isinstance(f, Callable):
        reveal_type(f)  # (...) -> Unknown

    if callable(f):
        reveal_type(f)  # Top[(...) -> object]
```
https://play.ty.dev/1081aeaf-3424-49f0-8e55-799e6544fcd5

The same is true for other checks, for example:

```py
from inspect import isawaitable
from typing import reveal_type, Awaitable

def _(f: object):
    if isinstance(f, Awaitable):
        reveal_type(f)  # Awaitable[Unknown]

    if isawaitable(f):
        reveal_type(f)  # Awaitable[object]
```

One could argue that this is a good thing: The behavior is not too hard to explain (`isinstance` checks are always gradual, stdlib `TypeIs` checks are always strict), and this gives users a choice to do both gradual and strict checks.

On the other hand, it would probably be more consistent to toggle the behavior of `callable`/`isawaitable`/etc with the `strict-generic-narrowing` option in the same way that the corresponding `isinstance` checks change behavior.

**Comment by alexprengere** [comment]
I was about to open an issue to report this. Currently I prefer having `callable` for runtime checks, but currently only the instance checks work as expected. Here is a basic MRE:

```python
from typing import Any, Callable

weight: float | Callable[[Any], float] = .5

if callable(weight):
    weight()  # fails with Object of type `Top[(...) -> object]` is not safe to call
```
Changing this to `isinstance(weight, Callable)` fixes the typing issue, as expected.

**Comment by AlexWaygood** [comment]
ty here is warning you that just because `weight` is callable doesn't mean that you know what the signature is. It could be a callable subclass of `float`, with an unknown signature. Strictly speaking our behaviour for `callable(weight)` is therefore more accurate than our behaviour for `isinstance(weight, Callable)`, and the safest thing to do here would be to use `if not isinstance(weight, float)` rather than `if callable(weight)` or `if isinstance(weight, Callable)`.

**Comment by carljm** [comment]
I think it is nice for users to have the option of both kinds of narrowing, but providing that option via the distinction between `if callable(x): ...` and `if isinstance(x, Callable): ...` is not intuitive and just looks like a bug. So I feel like ideally we would fix this and have the behavior of both be triggered by the strict-narrowing option.

**Comment by dsfaccini** [comment]
> This comment was posted by Claude Code using claude-fable-5 on behalf of David.

An ecosystem datapoint for the `isawaitable` half of this question, from running ty 0.0.72 over [pydantic-ai](https://github.com/pydantic/pydantic-ai).

The shape that actually bites us isn't narrowing from `object` — it's narrowing a **declared union** `X | Awaitable[X]`, i.e. the "user-supplied callback may be sync or async" idiom. There the strict `TypeIs` result collapses the awaited value all the way to `object`, while the `isinstance` spelling of the same check is fine:

```py
import inspect
from collections.abc import Awaitable


async def with_isawaitable(value: list[str] | Awaitable[list[str]]) -> list[str]:
    if inspect.isawaitable(value):
        return await value
    return value


async def with_isinstance(value: list[str] | Awaitable[list[str]]) -> list[str]:
    if isinstance(value, Awaitable):
        return await value
    return value
```

```console
$ uvx ty check repro.py
repro.py:7:16: error[invalid-return-type] Return type does not match returned value: expected `list[str]`, found `object`
Found 1 diagnostic
```

pyright and mypy are both clean on this file.

This is the same shape as #1567, which was closed not-planned with the advice to write `isinstance(obj, C)` instead — but as noted in that thread, that workaround doesn't exist when `X` is a generic alias like `list[str]`, and `@final` isn't available either. So today the only resolution for this pattern is a suppression comment.

It's not a rare shape for us: 13 diagnostics on 0.0.72, concentrated in the newest parts of the codebase (`pydantic_ai/capabilities/`, `pydantic_ai/toolsets/`), e.g. [`capabilities/prepare_tools.py`](https://github.com/pydantic/pydantic-ai/blob/main/pydantic_ai_slim/pydantic_ai/capabilities/prepare_tools.py) where a user hook typed `Callable[..., list[ToolDefinition] | Awaitable[list[ToolDefinition]]]` is awaited behind `inspect.isawaitable`.

Concrete question on the proposal in the issue body: if `isawaitable`/`callable` get toggled by `strict-generic-narrowing`, would that also relax this union case — where the pre-narrowing type already names the awaitable's type argument — or only narrowing from `object`/`Unknown`?

Notes: ty 0.0.72 via `uvx ty check`.

**Comment by carljm** [comment]
> Concrete question on the proposal in the issue body: if `isawaitable`/`callable` get toggled by `strict-generic-narrowing`, would that also relax this union case — where the pre-narrowing type already names the awaitable's type argument — or only narrowing from `object`/`Unknown`?

It would eliminate the diagnostic in your case also. (Assuming you did not turn on `strict-generic-narrowing`).

**Comment by sharkdp** [comment]
This has now been fixed in https://github.com/astral-sh/ruff/pull/28193.

The new behavior in non-strict mode is this:
```py
def _(f: object):
    if isinstance(f, Callable):
        reveal_type(f)  # (...) -> Unknown

    if callable(f):
        reveal_type(f)  # (...) -> object
```
The remaining inconsistency in the return type comes from typeshed's annotation.

In strict mode, both narrowing checks lead to the same result:
```py
def _(f: object):
    if isinstance(f, Callable):
        reveal_type(f)  # (...) -> Unknown

    if callable(f):
        reveal_type(f)  # (...) -> object
```
